src/parser.rs

```diff
@@ -551,175 +563,199 @@ pub enum SegmentTag {
-                SegmentTag::DateRange
-            )
-            | map!(ext_tag, SegmentTag::Unknown)
-            | map!(comment_tag, SegmentTag::Comment)
-            | map!(consume_line, SegmentTag::Uri)
-    )
-}
-
-named!(pub duration_title_tag<(f32, Option<String>)>,
-    do_parse!(
-           duration: float
-        >> opt!(tag!(","))
-        >> title: opt!(map_res!(is_not!("\r\n,"), from_utf8_slice))
-        >> take!(1)
-        >> opt!(tag!(","))
-        >>
-        (duration, title)
-    )
-);
-
-named!(pub key<Key>, map!(key_value_pairs, Key::from_hashmap));
-
-named!(pub extmap<Map>, map!(key_value_pairs, |attrs| Map {
-    uri: attrs.get("URI").cloned().unwrap_or_default(),
-    byte_range: attrs.get("BYTERANGE").map(|range| {
-        match byte_range_val(range.as_bytes()) {
-            IResult::Ok((_, br)) => br,
-            _ => panic!("Should not happen"),
-        }
-    }),
-}));
+pub fn media_segment_tag(i: &[u8]) -> IResult<&[u8], SegmentTag> {
+    alt((
+        map(
+            pair(tag("#EXTINF:"), duration_title_tag),
+            |(_, (duration, title))| SegmentTag::Extinf(duration, title),
+        ),
+        map(
+            pair(tag("#EXT-X-BYTERANGE:"), byte_range_val),
+            |(_, range)| SegmentTag::ByteRange(range),
+        ),
+        map(tag("#EXT-X-DISCONTINUITY"), |_| SegmentTag::Discontinuity),
+        map(pair(tag("#EXT-X-KEY:"), key), |(_, key)| {
+            SegmentTag::Key(key)
+        }),
+        map(pair(tag("#EXT-X-MAP:"), extmap), |(_, map)| {
+            SegmentTag::Map(map)
+        }),
+        map(
+            pair(tag("#EXT-X-PROGRAM-DATE-TIME:"), consume_line),
+            |(_, line)| SegmentTag::ProgramDateTime(line),
+        ),
+        map(
+            pair(tag("#EXT-X-DATE-RANGE:"), consume_line),
+            |(_, line)| SegmentTag::DateRange(line),
+        ),
+        map(ext_tag, SegmentTag::Unknown),
+        map(comment_tag, SegmentTag::Comment),
+        map(consume_line, SegmentTag::Uri),
+    ))(i)
+}
+
+pub fn duration_title_tag(i: &[u8]) -> IResult<&[u8], (f32, Option<String>)> {
+    map(
+        tuple((
+            float,
+            opt(char(',')),
+            opt(map_res(is_not("\r\n,"), from_utf8_slice)),
+            take(1usize),
+            opt(char(',')),
+        )),
+        |(duration, _, title, _, _)| (duration, title),
+    )(i)
+}
+
+pub fn key(i: &[u8]) -> IResult<&[u8], Key> {
+    map(key_value_pairs, Key::from_hashmap)(i)
+}
+
+pub fn extmap(i: &[u8]) -> IResult<&[u8], Map> {
+    map_res(key_value_pairs, |attrs| -> Result<Map, &str> {
+        let uri = attrs.get("URI").cloned().unwrap_or_default();
+        let byte_range = attrs
+            .get("BYTERANGE")
+            .map(|range| match byte_range_val(range.as_bytes()) {
+                IResult::Ok((_, range)) => Ok(range),
+                IResult::Err(_) => Err("invalid byte range"),
+            })
+            .transpose()?;
```

Comment:
Cool to see transpose here, didn't know rust had something like that by default. Now if only we could combine map + transpose into something like `traverse` or something 😄